### PR TITLE
Hive partitioned(v4) rebased

### PR DESCRIPTION
**Pull request by jpdna**
The "hive style" partitioned PR rebased against current ADAM master.

It will only compile against Spark2, hence POMs have been moved to Spark 2.
Is there a specific strategy wrt to handelling that for Spark2 with a profile you suggest?  model after how the R code works?

Please let me know, especially if you have comments/changes wrt to the partitioning related code.

todo:
1) its looks like there is some cleanup related to adam-cli/src/test/resources/artificial.fa that I need to look at
2) a few small formatting issues in ADAMContext - will fix so that this PR does not modify where its arbitrary

### CODE REVIEW

**Comment by heuermh** [comment]
> Is there a specific strategy wrt to handelling that for Spark2 with a profile you suggest?

Just wait until #1861 lands.

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/AlignmentRecordRDD.scala [hunk 1]
```diff
@@ -216,7 +216,7 @@ case class ParquetUnboundAlignmentRecordRDD private[rdd] (
     sc.loadParquet(parquetFilename)
   }
 
-  lazy val dataset = {
+  lazy val dataset: Dataset[AlignmentRecordProduct] = {
```
Comment (akmorrow13) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/AlignmentRecordRDD.scala]: Is this needed?

File: adam-cli/src/test/resources/artificial.fa [hunk 2]
```diff
@@ -1 +0,0 @@
-../../../../adam-core/src/test/resources/artificial.fa
```
Comment (akmorrow13) [inline adam-cli/src/test/resources/artificial.fa]: Why were there changes to this file?

File: adam-core/pom.xml [hunk 3]
```diff
@@ -211,7 +211,7 @@
     <dependency>
       <groupId>org.scala-lang</groupId>
       <artifactId>scala-library</artifactId>
-      <scope>provided</scope>
+      <scope>compile</scope>
```
Comment (akmorrow13) [inline adam-core/pom.xml]: What is this change for?

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMContext.scala [hunk 4]
```diff
@@ -1861,6 +1858,40 @@ class ADAMContext(@transient val sc: SparkContext) extends Serializable with Log
...
+   *
+   * @note The sequence dictionary is read from an Avro file stored at
+   *   pathName/_seqdict.avro and the record group dictionary is read from an
+   *   Avro file stored at pathName/_rgdict.avro. These files are pure Avro,
+   *   not Parquet + Avro.
+   *
+   * @param pathName The path name to load alignment records from.
+   *   Globs/directories are supported.
+   * @param regions
+   * @return
```
Comment (akmorrow13) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMContext.scala]: Descriptions of param and return types

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMContext.scala [hunk 5]
```diff
@@ -2244,6 +2275,33 @@ class ADAMContext(@transient val sc: SparkContext) extends Serializable with Log
     }
   }
 
+  /**
+   * Load a path name with range binned partitioned Parquet + Avro format into GenotypeRDD
+   *
+   * @param pathName
+   * @param regions
+   * @return
```
Comment (akmorrow13) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMContext.scala]: Description of param and return types

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMContext.scala [hunk 6]
```diff
@@ -3046,4 +3173,49 @@ class ADAMContext(@transient val sc: SparkContext) extends Serializable with Log
       loadParquetFragments(pathName, optPredicate = optPredicate, optProjection = optProjection)
     }
   }
+
+  /**
+   * Test if Parquet files are partitioned
+   *
+   * @param filePath
+   * @return
```
Comment (akmorrow13) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMContext.scala]: param and return type description

**Comment by jpdna** [comment]
Question about rebase/history:
So - after having some troubles resolving conflicts during rebasing, I ended up using the "resolve" tool on github to merge master into my branch.
This seems to have worked in terms of the final product is what I want with only intended changes on top of Master.

HOWEVER - I don't like the way that the history has worked out with, with minor incremental changes which I would have liked to have squashed interleaved with the commits on Master.

@fnothaft @heuermh  or others can you advise me on if there is  a need - and what is the best way, to ideally go from where this PR branch is now, to a cleaner history?   

Here is what I am considering: Should I make a new branch from master, compare it to this PR branch, then move the intended changes over manually to that new branch, to consolidate all the changes into a single commit on top of current Master?

**Comment by jpdna** [comment]
Closed in favor a clean single commit for these changes at: #1878
The comments on this issue I believe have been addressed.

**Pull request closed without merging**